sips/sip-010/sip-010-fungible-token-standard.md

```diff
@@ -0,0 +1,178 @@
+# Preamble
+
+Sip Number: 010
+
+Title: Standard Trait Definition for Fungible Tokens
+
+Author: Hank Stoever (hstove@gmail.com)
+
+Consideration: Technical
+
+Type: Standard
+
+Status: Draft
+
+Created: 25 January 2021
+
+License: CC0-1.0
+
+Sign-off:
+
+Layer: Traits
+
+Discussions-To: https://github.com/stacksgov/sips
+
+# Abstract
+
+Fungible tokens are digital assets that can be sent, received, combined, and divided. Most forms of cryptocurrencies are fungible tokens. They have become a building block of almost all blockchains. This SIP aims to provide a flexible and easy-to-implement standard that can be used by developers on the Stacks blockchain when creating their own tokens.
+
+# License and Copyright
+
+This SIP is made available under the terms of the Creative Commons CC0 1.0 Universal license, available at https://creativecommons.org/publicdomain/zero/1.0/
+This SIP’s copyright is held by the Stacks Open Internet Foundation.
+
+# Introduction
+
+Digital assets can have the property to be fungible. A _fungible_ token can be broken down into small units and added together. An owner of a fungible asset only needs to care about their balance, that is, the total amount of a particular fungible asset that they own. Most well-known currencies are fungible. For fungible tokens, there is no difference between any two different amounts of the fungible token.
+
+For example, if a user owns 10 units of a fungible asset, they may send 2 units to a different user. At this point, their balance is 8 units. If they later receive more units, their total balance will be updated.
+
+On blockchains, fungible tokens are a core component. Blockchains with smart contracts, including the Stacks blockchain, allow developers and users to create and interact with smart contracts that use fungible tokens.
+
+The Stacks blockchain has a native fungible token: the Stacks token (STX). In addition to the native STX token, the Stacks blockchain's programming language for developing smart contracts, Clarity, has built-in language primitives to define and use fungible tokens. Although those primitives exists, there is value in defining a common interface (known in Clarity as a "trait") that allows different smart contracts to interoperate with fungible token contracts in a reusable way. This SIP defines that trait.
+
+# Specification
+
+The fungible token trait, `ft-trait`, has a 6 functions:
+
+## Trait functions
+
+### Transfer
+
+`(transfer ((amount uint) (sender principal) (recipient principal)) (response bool uint))`
```

Comment:
This is something that came up in conversation with @lgalabru: whether or not we should recommend something akin to _allowances_ in the `ft-trait` standard. With ERC-20 allowances, senders can specify that a given smart-contract is allowed to transfer up to some amount on their behalf _without_ needing to be called by a user.

For example, a "decentralized exchange" could work like so:

1. Alice invokes `(contract-call? stackaroos-contract grant-allowance dex-contract 1000)`
2. Alice invokes `(contract-call? dex-contract trade-order stackaroos-contract 1000 stackoalas 500) -> returns order number 1`
3. Bob executes the order `(contract-call? dex-contract execute-order 1)`

And the transfers are all executed at (3). (1) could even be invoked by the implementation of (2), such that all this happens in 2 transactions, without the stackaroos leaving Alice's wallet until the order is executed.

Without allowances, Alice would need to transfer the stackaroos to the `dex-contract` and then reclaim them after the order expires (or revoke the order).

I'm not sure which approach is the best experience for users. If we went with a design that used allowances, we'd want to recommend that the allowance functions mint an "allowance" FT, which would allow the transaction caller to use post-conditions to protect themselves from unknowingly creating FT allowances.

Reply:
Could we allow both?  Allowances could be a separate trait that the contract could implement.  Even if a token gets implemented with this trait, I could imagine someone writing a proxy contract for it that implements both the ft-trait and the allowance trait simply by wrapping the ft-trait methods and implementing the allowance mechanism.

Reply:
Going to go ahead and resolve this.  If later the community decides they want fungible tokens to have allowances, then it can be done in a separate trait (with a separate SIP).